### ISSUE TITLE
Wingpanel does not reserve space

### ISSUE THREAD

**Issue by Miosame** [elementary/wingpanel#315]
## Prerequisites
- [x] I have searched open and closed issues for duplicates.

## Describe the bug
If I maximize a window without it going actual "fullscreen" (F11 / win+F) - it goes under the wingpanel, making it seem as though wingpanel does not reserve space for itself and allows the window that maximizes to go under it.

In my VM this bug does not occur, but on my full setup (nvidia 1050ti + 4 monitors) it does. Note that this is a fresh untouched install.

If it helps, that's my screen configuration (if maybe the top monitor is messing with the reserved space?) - the one in the middle is set as my main (starred) screen.

![image](https://user-images.githubusercontent.com/8201077/80555510-bde00700-89d0-11ea-8a78-dc61f3768721.png)

## Platform Information

 - OS: elementary OS
 - OS Version 5.1.3 Hera
 - Hardware info: nvidia 1050ti

---

- [ ] I'm using the latest version from git that I've manually compiled
- [x] I'm using the latest released stable version

**Comment by cassidyjames** [comment]
@Miosame thanks for reporting this! We recently merged [some fixes](https://github.com/elementary/wingpanel/pull/295) for three-display layouts (thanks @andreasfelix!), but I wonder if we could do something to address this layout as well. :thinking: If we come up with something to test, it would be great if we could walk you through that process and get your feedback—I'm not sure anyone on the team has access to four physical displays, especially since many are sheltering-in-place at the moment.

**Comment by Miosame** [comment]
@cassidyjames the odd thing to me is that even if I extend vmware across all my 4 displays, via its inbuilt function:

![image](https://user-images.githubusercontent.com/8201077/80557163-5036d980-89d6-11ea-8b66-c784df46520a.png)

it maximizes just fine, it only happens on metal / non VM - nvidia backed hardware install.

---

Did the mentioned changes get merged already to be usable in elementary 5.1.3 or do I have to compile it? if the latter, do I just build it as stated in the readme and execute wingpanel or will that cause some issues? aka do I have to first purge the old wingpanel?

**Comment by Miosame** [comment]
Looking at the recent commits - I assume it's not safe to clone master and compile? as it did not pass pipeline. What commit would be safe to try for this issue?

**Comment by Miosame** [comment]
Compiling it all on the real machine and restarting did not fix the issue sadly, is there more steps I should do or are the error messages above what makes it not compile right?

**Comment by andreasfelix** [comment]
To my understanding, this is a corner case which can not be fixed without a new [freedesktop.org]( https://www.freedesktop.org/wiki/) specification and an implementation in the window manager.

If you look a the code

https://github.com/elementary/wingpanel/blob/1b2257d0f477b9eab660e59057a551dd8ef33f12/src/PanelWindow.vala#L188-L189

you can see that the wingpanel uses the [`_NET_WM_STRUT_PARTIAL` ](https://specifications.freedesktop.org/wm-spec/wm-spec-1.5.html#NETWMSTRUT) specification to reserve its space. The problem is that, the `_NET_WM_STRUCT_PARTICAL` specification is limited in that way that it does not allow to reserve space for arbitrary rectangles. Instead it ONLY allows to reserve space at the borders of screen area.

I made a visualization to show what this means:

![multi_monitor](https://user-images.githubusercontent.com/24651767/62983252-fc2b9b00-be2e-11e9-86a1-21b440cf1bc0.png)

*Solid dark lines: Monitors, Blue area: area that client reserves. Dotted area: screen area*

To reserve space the blue rectangle has to start from the dotted area. For setups where the primary monitor has a left, right and top neighbor (example with ❌), I think there is no way one can draw such an rectangle without interfering with other monitors.

**Comment by andreasfelix** [comment]
> the odd thing to me is that even if I extend vmware across all my 4 displays, via its inbuilt function:
> 
> ![image](https://user-images.githubusercontent.com/8201077/80557163-5036d980-89d6-11ea-8b66-c784df46520a.png)
> 
> it maximizes just fine, it only happens on metal / non VM - nvidia backed hardware install.

I assume in your VM, you only have one monitor. In this case the wingpanel is able to reserve the space, because it is effectively a single monitor setup. So there are no other monitors screening the wingpanel from the edge of the screen area.

**Comment by Miosame** [comment]
Thanks for the visualization! though now it makes me wonder how other gtk based environments handle this? see Ubuntu 20.04 for example which does have a very similar top bar, which does reserve space on this same layout.

Would it be maybe beneficial to look at their implementation? I'm sadly not great at anything GTK related, so all I can do is help testing builds.

Regarding the VM, I'll check shortly the display configuration it exposes to the guest, probably it indeed is one big monitor but with boundaries set virtually by VMware somehow.

**Comment by Miosame** [comment]
@andreasfelix here's the VM check:

![image](https://user-images.githubusercontent.com/8201077/80620582-77c58a80-8a46-11ea-9286-411980af07f8.png)

it actually does recognize it in the same config as my "real" setup, it's not a single giant virtual monitor.

**Comment by andreasfelix** [comment]
In the VM it work because you have the windows aligned at the bottom.

**Comment by andreasfelix** [comment]
Please show a screenshot of your actual machine

**Comment by Miosame** [comment]
Can confirm, if I move in my real machine the side displays to the bottom, the reserving works again :smile: - would it still be worth it to check how Ubuntu does it? since there even if the side monitors are aligning the top it works.

I'd be glad to test anything you want to try for this too. :grinning:

**Comment by Miosame** [comment]
Oh another thing I noticed, building wingpanel from scratch, it doesn't get themed anymore as the original one does when gtk is set to prefer dark themes - is that a known issue of compiling from scratch or should I file an issue?

**Comment by andreasfelix** [comment]
I looked into the GNOME/Ubuntu code. 

For GNOME3, my understanding is that the panel is built into the GNOME Shell itself and the GNOME Shell just internally reduces the workarea by the size of the panel:

https://github.com/GNOME/gnome-shell/blob/867587ef4cb92a2f42580de8ec508198d43f4044/js/ui/layout.js#L1091-L1097

For the older GNOME/Ubuntu Panel (Flashback) I found something interesting:

https://github.com/GNOME/gnome-panel/blob/135c13fb743f9c7bfdbd3e666ef59feae82b5693/gnome-panel/panel-xutils.c#L72-L77

It uses the `_GNOME_WM_STRUT_AREA` specification which can be used to reserve rectangles starting from individual monitors instead of the entire screen. But the specification seems to be relatively new and has to be implemented in mutter/gala to work on elementary OS.

See also this [blogpost](https://muktupavels.id.lv/posts/better-multi-monitor-support-in-gnome-flashback).

Regarding your theming issue: I am also just a occasional contributor and don't know about that.

@cassidyjames  At the end of the  [GNOME Gitlab discussion](https://gitlab.gnome.org/GNOME/mutter/-/issues/452#note_706118) Alberts Muktupāvels offered to implement `_GNOME_WM_STRUT_AREA` specification in mutter. Is this something elementary would be interested in?

**Comment by cassidyjames** [comment]
@andreasfelix absolutely, that sounds like it would work better for us. Thanks for looking into this!

**Comment by cassidyjames** [comment]
@Miosame there's not currently an official dark style support in elementary OS, but I would guess our recent behind-the-scenes efforts in that direction may be conflicting with however you were forcing it. Soon you should be able to set a more official hidden setting while we work towards supporting it throughout the OS.

**Comment by Miosame** [comment]
@cassidyjames I was using elementary-tweaks that sets "prefer dark theme" before

**Comment by cassidyjames** [comment]
@Miosame yeah, and I think that's not compatible with proper dark style support—it works by hackily globally overriding an application setting for all apps. The way proper dark style support works is by checking a user-set preference on the account and then setting the GTK dark style variant based on that preference. Since WingPanel has been updated to check for this user preference if it exists, it's possible it's overriding the global hack.
